### PR TITLE
[P2-M01] Subject transfers of sponsor positions to liveness period

### PR DESCRIPTION
**Pull request by nicholaspai**
The goal of this PR is to prevent sponsors from delaying liquidations by transferring their position to another address that they control before a liquidation transaction is mined. With instantaneous transfers (as they are now), the sponsor could delay liquidations indefinitely.

This PR modifies the `transfer` method to `requestTransfer`/`cancelTransfer`/`transferPassedRequest` similar to the `withdraw` pattern. To prevent adding another liveness parameter to the constructor configuration, I re-use the `withdrawLiveness` parameter. N.B. I think that the `liquidationLiveness` would be a more ideal choice here because the main exploit in `transfer` is to delay liquidations, but the `PricelessPositionManager` does not currently have access to the `liquidationLiveness` variable.

One thing to note: When the user requests a transfer, they do not input which address to transfer the position to. I chose this because there wasn't a natural place to store this address outside of creating a new mapping for pending transfers. Moreover, I didn't see any reason why the sponsor could not specify the `newSponsor` address post-liveness.

This PR has a few follow-ons we should consider
- @chrismaree has concluded that the EMP is near its bytecode limit and this would increase the bytecode size by ~500 bytes
- @ptare this will require a follow-up PR to modify the `transfer` flow in the sponsor CLI scaffolding

Signed-off-by: Nick Pai <npai.nyc@gmail.com>

### CODE REVIEW

**Comment by mrice32** [comment]
@ptare could you take a pass on this?

**Comment by ptare** [comment]
IMO it's too confusing to reuse `requestPassTimestamp` for this purpose. Are we sure that there is no inadvertent interaction between withdrawals and transfers?

**Comment by nicholaspai** [comment]
> IMO it's too confusing to reuse `requestPassTimestamp` for this purpose. Are we sure that there is no inadvertent interaction between withdrawals and transfers?

This is a fair point. I think its possible that you could `requestTransfer` and then call `withdrawPassedRequest`. To be cleaner, I'll just add additional data to the `PositionData` struct.

**Comment by nicholaspai** [comment]
@chrismaree @mrice32 @ptare do you think we should freeze all actions whenever there is a pending `transfer`? Similar to how we freeze all actions when there is a pending withdrawal request?

The reason we freeze the modification of collateral amounts during a pending withdrawal is to allow liquidations to occur on the pending withdrawal amount. Since a `transfer` does not change collateral, my initial thought is to not block anything.

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 1]
```diff
@@ -47,6 +47,8 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
         // Raw collateral value. This value should never be accessed directly -- always use _getCollateral().
         // To add or remove collateral, use _addCollateral() and _removeCollateral().
         FixedPoint.Unsigned rawCollateral;
+        // Tracks pending transfer requests. A transfer request is pending if `transferRequestPastTimestamp != 0`.
+        uint256 transferRequestPassTimestamp;
```
Comment (nicholaspai) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol:51]: @ptare added this data here to track pending transfers

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 2]
```diff
@@ -165,21 +169,62 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
...
+    }
+
+    /**
+     * @notice After a passed transfer request (i.e., by a call to `requestTransfer` and waiting
+     * `withdrawalLiveness`), transfers ownership of the caller's current position to `newSponsorAddress`.
      * @dev Transferring positions can only occur if the recipient does not already have a position.
      * @param newSponsorAddress is the address to which the position will be transferred.
      */
-    function transfer(address newSponsorAddress) public onlyPreExpiration() {
+    function transferPassedRequest(address newSponsorAddress) public onlyPreExpiration() {
```
Comment (nicholaspai) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol:196]: Juts want to highlight again that I made a design decision to allow the sponsor to specify the `newSponsorAddress` on this call as opposed to when they initially make the `requestTransfer` call. I saw no downside to giving sponsor flexibility post-liveness and the upside is we don't have to store an `address` in `PositionData`.

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 3]
```diff
@@ -165,21 +169,62 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
...
         require(_getCollateral(positions[newSponsorAddress].rawCollateral).isEqual(FixedPoint.fromUnscaledUint(0)));
         PositionData storage positionData = _getPositionData(msg.sender);
         require(positionData.requestPassTimestamp == 0);
```
Comment (nicholaspai) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol:199]: Note that `transferPassedRequest` is blocked whenever there is a withdrawal request pending, but neither `requestTransfer` nor `cancelTransfer` are blocked.

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 4]
```diff
@@ -165,21 +169,62 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
      ****************************************/
 
     /**
-     * @notice Transfers ownership of the caller's current position to `newSponsorAddress`.
+     * @notice Requests to transfer ownership of the caller's current position to `newSponsorAddress`. Once the request liveness is passed,
+     * the sponsor can execute the transfer.
+     * @dev The liveness length is the same as the withdrawal liveness.
+     */
+    function requestTransfer() public onlyPreExpiration() {
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply core/contracts/financial-templates/implementation/PricelessPositionManager.sol:176]: yup exactly. But no reason to block pending transfers while a withdrawal is in progress.

**Pull request merged**